libp2p/src/builder/phase/dns.rs

```diff
@@ -48,6 +48,49 @@ impl<T: AuthenticatedMultiplexedTransport> SwarmBuilder<super::provider::Tokio,
     }
 }
 
+#[cfg(all(not(target_arch = "wasm32"), feature = "async-std", feature = "dns"))]
+impl<T: AuthenticatedMultiplexedTransport> SwarmBuilder<super::provider::AsyncStd, DnsPhase<T>> {
+    pub async fn with_dns_config(
+        self,
+        cfg: libp2p_dns::ResolverConfig,
+        opts: libp2p_dns::ResolverOpts,
+    ) -> SwarmBuilder<
```

Comment:
For other reviewers:

- Our `with_tcp` always takes a config. There is no `with_tcp_config`.
- Our `with_quic_config` takes a closure. This allows the builder to inject the keypair before passing the config to the user to be further configured.
    https://github.com/libp2p/rust-libp2p/blob/4e1ad09c779ba76c0cd6a672981ca780cb0c277c/libp2p/src/builder/phase/quic.rs#L32

This inconsistency is not ideal. I suggest we go with `with_dns_config` here. I suggest the method to take the `ResolverConfig` and `ResolverOpts` directly, as implemented here. Neither needs the keypair and thus a closure isn't needed.

Reply:
> This inconsistency is not ideal.

The best one to resolve would probably be to have `with_tcp` take a closure that allows you to configure the TCP transport better.